### PR TITLE
Adds areagraph tooltips

### PR DESCRIPTION
**Pull request by Kongkille**
_Can be reviewed per commit_

## Issue
Chipping away at https://linear.app/electricitymaps/issue/ELE-1342/[ui]-panel-zone-details-charts.

## Description

This uses somewhat of a [render prop pattern](https://reactjs.org/docs/render-props.html) to be able to handle the positioning and being able to switch out different tooltips without duplicating code.

I'll follow this up with creating the actual tooltips

FYI @madsnedergaard the idea is that the bar breakdown chart should use the same 'BreakdownChartTooltip' but it will likely not be a child of the areaGraphTooltip

### Preview

https://user-images.githubusercontent.com/45588799/205905238-e23fd68d-52b9-434d-8a29-081edeb9428f.mov

### CODE REVIEW

File: web/src/components/tooltips/AreaGraphTooltip.tsx [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+): ReactElement | null {
+  const { children, zoneDetail, selectedLayerKey, position } = properties;
+
+  if (
+    children === undefined ||
+    selectedLayerKey === undefined ||
+    zoneDetail === undefined
+  ) {
+    return null;
+  }
```
Comment (Kongkille) [inline web/src/components/tooltips/AreaGraphTooltip.tsx:33]: I don't like this. Do you have any idea on how to not check all the things for undefined?

I can of course check before rendering but then I'd have the same check just in another place.

File: web/src/components/tooltips/AreaGraphTooltip.tsx [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+  const screenWidth = window.innerWidth;
```
Comment (madsnedergaard) [inline web/src/components/tooltips/AreaGraphTooltip.tsx:35]: Do we need to use the resize observer thing here?

**Comment by Kongkille** [comment]
I forgot to branch out and merge this so now everything is in this PR. If it gets too large then lmk and i'll split


https://user-images.githubusercontent.com/45588799/207548998-3ff28f61-8d97-4038-99ec-08b464bfd9d1.mov

File: web/src/features/panels/LeftPanel.tsx [hunk 3]
```diff
@@ -44,7 +44,7 @@ function OuterPanel({ children }: { children: React.ReactNode }) {
         !isOpen && '-translate-x-full'
       }`}
     >
-      <section className="w-full p-2">{children}</section>
+      <section className="w-full py-2 pl-1 pr-2">{children}</section>
```
Comment (Kongkille) [inline web/src/features/panels/LeftPanel.tsx:47]: This makes the left spacing not too large

Comment (madsnedergaard) [reply web/src/features/panels/LeftPanel.tsx:47]: How about just overriding the specific side instead?
```suggestion
      <section className="w-full p-2 pl-1">{children}</section>
```

**Comment by madsnedergaard** [comment]
> I forgot to branch out and merge this so now everything is in this PR. If it gets too large then lmk and i'll split
> 
>  Screen.Recording.2022-12-14.at.09.46.40.mov

Why is the tooltip staying after you leave the graph, shouldn't it close/disappear immediately after leaving?

**Comment by Kongkille** [comment]
> > I forgot to branch out and merge this so now everything is in this PR. If it gets too large then lmk and i'll split
> > Screen.Recording.2022-12-14.at.09.46.40.mov
> 
> Why is the tooltip staying after you leave the graph, shouldn't it close/disappear immediately after cursor leaves?

Ah yeah ok, this is because of the recording, there was a popup, so it messed up the UX lol, it does leave correctly otherwise

File: web/src/components/tooltips/AreaGraphTooltip.tsx [hunk 4]
```diff
@@ -32,13 +30,15 @@ export default function AreaGraphTooltip(
     return null;
   }
 
+  console.log(tooltipSize);
+
   const screenWidth = window.innerWidth;
   const tooltipWithDataPositon = getTooltipPosition(
```
[the remark this answers is not in the document]
Comment (Kongkille) [reply web/src/components/tooltips/AreaGraphTooltip.tsx]: agreed

File: web/src/components/tooltips/AreaGraphTooltip.tsx [hunk 5]
```diff
@@ -0,0 +1,51 @@
+import * as Portal from '@radix-ui/react-portal';
```
[the remark this answers is not in the document]
Comment (Kongkille) [reply web/src/components/tooltips/AreaGraphTooltip.tsx:1]: we can do that

File: web/public/locales/ar.json [hunk 6]
```diff
@@ -14,8 +14,13 @@
     "emissions": "انبعاثات الكربون",
     "addeditsource": "<a href=\"%s\" target=\"_blank\">أضف أو عدّل على المصدر</a>"
   },
-  "country-history": { "carbonintensity": { "hourly": "كثافة الكربون في آخر 24 ساعة مضت" } },
-  "footer": { "foundbugs": "وجدت أخطاء برمجية أو لديك اقتراحات؟ أبلغنا عنها", "here": "هنا" },
+  "country-history": {
+    "carbonintensity": { "hourly": "كثافة الكربون في آخر 24 ساعة مضت" }
```
[the remark this answers is not in the document]
Comment (Kongkille) [reply web/public/locales/ar.json:18]: Perhaps it is the new line width changes?

[the remark this answers is not in the document]
Comment (madsnedergaard) [reply web/public/locales/ar.json:18]: Ah looks fine now, I just thought the 24 number had been moved around, but that doesn't seem to be the case :D

**Comment by tonypls** [comment]
I can't seem to get this to run locally, when I add remote=true I get a spinner for ever 😢

**Comment by tonypls** [comment]
> I can't seem to get this to run locally, when I add remote=true I get a spinner for ever 😢

I take it back, even master is doing it for me, must be something on my machine. Maybe the ghost of 1000 wind layers

File: web/src/components/tooltips/utilities.ts [hunk 7]
```diff
@@ -0,0 +1,45 @@
+export const getSafeTooltipPosition = (
```
Comment (tonypls) [inline web/src/components/tooltips/utilities.ts:1]: This exudes safety :D

File: web/src/features/charts/tooltips/CarbonChartTooltip.tsx [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+
+  const co2ColorScale = useCo2ColorScale();
+
+  if (!zoneDetail) {
+    return null;
+  }
+  const { co2intensity, stateDatetime } = zoneDetail;
+
+  return (
+    <div className="w-[320px] rounded-md bg-white p-3 shadow-xl">
```
Comment (tonypls) [inline web/src/features/charts/tooltips/CarbonChartTooltip.tsx:23]: At some point we should make all the shadows match where logical

File: web/src/features/charts/tooltips/EmissionChartTooltip.tsx [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+    return null;
+  }
+
+  const totalEmissions =
+    Math.round(tonsPerHourToGramsPerMinute(getTotalElectricity(zoneDetail, true)) * 100) /
+    100;
+  const { stateDatetime } = zoneDetail;
+
+  return (
+    <div className="w-[350px] rounded-md bg-white p-3 shadow-xl">
```
Comment (tonypls) [inline web/src/features/charts/tooltips/EmissionChartTooltip.tsx:23]: I only recently realised how high the built in tailwind spacing goes, if it makes sense here w-80 = 320px or w-96 = 384px. If we start trying to stick the the provided spacing hopefully things will line up better

**Review by tonypls — APPROVED** [review]
Looks pretty good, a few comments. From a usability perspective, do we want the tooltip to flip to the left when going beyond the window on smaller screens?

**Comment by Kongkille** [comment]
This doesn't really work well on mobile but I'll merge it for now and then I can fix it in a subsequent PR

**Pull request merged**